strapi/strapi
# feat!: document service query param validation

### What does it do?

- allows for custom validators to be used instead of only having the content API defaults
- adds validation to every document service method so that it will only accept valid parameters for `filters`, `fields`, `sort`, and `populate`

Specifically, it also adds new validations (inc. for content API) for:
- populate: unpopulatable fields
- populate: full validation for nested dynamic zones and components

Not included in this PR but should be coming at a later date are `status` `locale` and `pagination` validation

### Why is it needed?

To improve DX by throwing errors when invalid parameters are passed in to document service calls.

In Typescript those errors should already be caught in build, but this adds another layer of protection at runtime and will prevent any unexpected behaviour

### How to test it?

Try making document service calls with invalid query params (filters, fields, sort, populate) such as non-existant fields, sorting or filtering on morph relations, etc.

Full test coverage is included, and some additional content API tests have been added to ensure we aren't accidentally allowing anything we were previously validating

### Related issue(s)/PR(s)

DX-861
